src/io/aleph/dirigiste/AtomicLongRingBuffer.java

```diff
@@ -0,0 +1,32 @@
+package io.aleph.dirigiste;
+
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicLongArray;
+
+public class AtomicLongRingBuffer {
+
+    private final AtomicLongArray _values;
+    private final int _size;
+    private volatile int _count;
+    private volatile int _offset;
+
+    public AtomicLongRingBuffer(final int size) {
+        _values = new AtomicLongArray(size);
+        _size = size;
+    }
+
+    public synchronized void add(final long value) {
+        _values.set(_offset++, value);
+        _count = Math.min(_count + 1, _size);
+        _offset %= _size;
+    }
+
+    public long[] toSortedArray() {
+        long[] vals = new long[_count];
+        for (int i = 0; i < _count; i++) {
+            vals[i] = _values.get(i);
+        }
+        Arrays.sort(vals);
+        return vals;
+    }
```

Comment:
Hmmm, how do we know this doesn't need to be synchronized, too?

Reply:
I think the failure mode is really gentle, though, since `_count` could be lower than it should be (if an early `add` has updated `_values`, but not `_count` yet), but never higher than the array's length. And once the ring fills up, the `_count` will never change, so we'll always get the full array.

Reply:
> Hmmm, how do we know this doesn't need to be synchronized, too?

We are not taking important decisions based on `_queueLatencies` and `_taskLatencies` queues.
I don't think it's a big deal if the value is off a little bit on the worst case scenario.

The really important one is `_utilizations` which is used only by one Thread and doesn't need to be synchronized.

I'm not against synchronizing if you think it's needed.

Reply:
Let's not, then. And if we do, let's sync around just the array copy, not sorting, a much slower operation.